### ISSUE TITLE
[ Core|Serve] Ray Serve inside Ray Remote introduces slowdowns related to number of available CPU's

### ISSUE THREAD

**Issue by avicooper1** [ray-project/ray#46362]
### What happened + What you expected to happen

I think there is a bug related to using `@serve.deployment` inside a remote function when `max_calls > 0`, where execution slows down dramatically after the number of executed tasks exceeds the number of CPU's. 

#### Scenario
I have several servers with pre-loaded LLM's. I want to query these servers inside a set of remote functions

#### Bug
Each task consumes CPU's, and does not seem to release them correctly after completing. If there are `n` CPU's on the cluster, only `n` tasks will be run. (And if each task requests `m` cpus, then only `n/m` tasks will be run.) After reaching this point, execution will continue but extremely slowly (like running 1 task after few seconds.) Further, the dashboard indicates that the CPU's are still available. For reference, here is the dashboard:

[
<img width="1584" alt="Screenshot 2024-07-01 at 2 42 08 PM" src="https://github.com/ray-project/ray/assets/11204241/f80a0afe-ec14-4daa-9e00-49fd44f695b3">
](url)

It seems like this is only an issue when `max_calls > 0`, (either set explicitly like in the example below, of when set implicitly, like when requesting GPU resources,) when also calling a serve deployment method.

I've done several experiments to confirm that this is an issue related to calling a method of a Ray Serve deployment, including replacing the `main` remote function with:

1. Simple python code, like adding numbers
2. A call to a simple remote function
3.  A call to a method of a Ray actor (but not a `@serve.deployment`)
4. Retrieving the handle with `handle = serve.get_app_handle(name="app1")`, but not calling `handle.remote().result()`

and the issue does not arise.

It seems whether calling `handle.remote().result()` or `handle.remote()` the issue arises. (With the latter however more tasks are executed before slowdown.)

Since this only occurs in the specific case of `max_calls > 0` when calling a serve deployment method, it seems like a bug to me, but I'm not sure. Any guidance would be greatly appreciated!

### Versions / Dependencies

Tested with a clean install, Python: 3.10.12, Ray: 2.31.0

### Reproduction script

```python
import time
import ray
from ray import serve

@serve.deployment
class Server:
    def __call__(self):
        time.sleep(0.1)

@ray.remote(max_calls=1)
def main():
    handle = serve.get_app_handle(name="app1")
    handle.remote().result()

if __name__ == "__main__":
    serve.run(Server.bind(), name="app1")
    futures = []
    for j in range(1000):
        futures.append(main.remote())
    ray.get(futures)
```

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

**Comment by jjyao** [comment]
`@ray.remote(max_calls=1)` this basically disables the worker process reuse and that's probably where the slowdown is from since now each remote task needs to wait for a new worker process to start. Is there a reason why you need to set `max_calls`?

**Comment by avicooper1** [comment]
Yes, I assumed that it was something related to the restarting of the processes. However it slows down much more than I would have expected for restarting the processes.

The example I included above is a minimal working example. However I intend to use gpu's, which will automatically set `max_calls` to 1. (I initially discovered the issue just by requesting gpu resources.)
